### PR TITLE
Code testing guidelines added

### PR DESCRIPTION
**Pull request by ghtmtt**
### Description
First draft on how to write code snippet tests. ref #3544

Goal:

Explain how to write good and testable python documentation.

Ticket: #3544

- [ ] Backport to LTR documentation is required

### Minimal requirements for merging *(for maintainers)*

- [x] The description of this PR is coherent with the manual and does not provide wrong information.
- [x] This PR passes the checks.

### CODE REVIEW

File: source/docs/documentation_guidelines/testing_guidelines.rst [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+.. note:: all the code within the ``.. testsetup:: *`` is not visible in the
+    final file.
+
+After this you have to use the ``.. testcode::`` directive followed by a name
+you want, like ``.. testcode:: example``. Within this directive you should add
+all the code you want (basically all the code you added in the *old* ``.. code-block:: python``
+directive).
+
+Now the core part. After you wrote the code you can add some *assertion* that
```
Comment (DelazJ) [inline source/docs/documentation_guidelines/testing_guidelines.rst]:
```suggestion
After you wrote the code you can add some *assertion* that
```
Sounds like all you wrote up to now was not interesting/important!

**Comment by SrNetoChan** [comment]
I started to make some reviews and suggestions but I ended up fetching your branch and suggesting a PR on it:

Please have a look.

https://github.com/ghtmtt/QGIS-Documentation/pull/1/files

**Comment by DelazJ** [comment]
Oups, I was in the file tab and di not notice your comment @SrNetoChan

File: source/docs/documentation_guidelines/testing_guidelines.rst [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+*********************
+
+If you are planning to add or review some chapters of the pyqgis cookbook, then
+you have to follow some rules to enable automatic testing of code snippets.
+
+Testing is really important because we can check automatically the quality of the
+code: if the code is not good, tests will fail meaning that the snippet has
+some errors.
+
+The framework follow the `Sphinx rules <https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html>`_:
```
Comment (havatv) [inline source/docs/documentation_guidelines/testing_guidelines.rst]:
```suggestion
The framework follows the `Sphinx rules <https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html>`_:
```

File: source/docs/documentation_guidelines/testing_guidelines.rst [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+.. code-block:: python
+
+    crs = QgsCoordinateReferenceSystem(4326, QgsCoordinateReferenceSystem.PostgisCrsId)
+    assert crs.isValid()
+
+Basically you are creating a crs and with ``assert crs.isValid()`` you **test**
+if it is valid.
+
+You can of course run this code also in the python console or in a standalone
```
Comment (havatv) [inline source/docs/documentation_guidelines/testing_guidelines.rst]:
```suggestion
You can of course run this code also in the Python console or in a standalone
```

File: source/docs/documentation_guidelines/testing_guidelines.rst [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+script, but the automatic test engine will take care to test the snippet and
```
Comment (havatv) [inline source/docs/documentation_guidelines/testing_guidelines.rst]:
```suggestion
script, but the automatic test engine will test the snippet and
```

**Comment by ghtmtt** [comment]
thanks guys for the review.. @havatv I don't know your standard workflow, but could you make a single review with all the comments/suggestions? Else I get my email spammed :)

File: source/docs/documentation_guidelines/index.rst [hunk 5]
```diff
@@ -11,3 +11,4 @@ Documentation Guidelines
    do_translations
    substitutions
    first_contribution
+   testing_guidelines
```
[the remark this answers is not in the document]
Comment (ghtmtt) [reply source/docs/documentation_guidelines/index.rst]: I agree to rename it but not on moving it. I really think this should stay at the end: I'm pretty sure just a few people will work on that and therefore we should keep the `first_contribution` first

File: source/docs/documentation_guidelines/testing_guidelines.rst [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+``assert object.isValid()``) you can still compare the output with another
+directive: ``.. testoutput:: mysnippet``.
+
+Basically you have first to create your code with a known name, e.g. `.. testcode:: mysnippet``
+that contains some ``print`` statements:
+
+.. code-block:: python
+
+    print("QGIS CRS ID:", crs.srsid())
+    print("PostGIS SRID:", crs.srid())
```
[the remark this answers is not in the document]
Comment (ghtmtt) [reply source/docs/documentation_guidelines/testing_guidelines.rst]: why removing this?

File: source/docs/documentation_guidelines/testing_guidelines.rst [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+and then adding the expected results within the ``.. testoutput:: mysnippet``::
+
+    QGIS CRS ID: 3452
```
[the remark this answers is not in the document]
Comment (ghtmtt) [reply source/docs/documentation_guidelines/testing_guidelines.rst]: see above (why removing 2 valid tests?)

**Comment by SrNetoChan** [comment]
@ghtmtt can you please check my PR, I have shuffled stuff around a bit.

File: source/docs/documentation_guidelines/testing_guidelines.rst [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    PostGIS SRID: 4326
+
+
+How to test snippets on your local machine
+==========================================
+
+.. note:: instructions are valid for Linux system.
```
[the remark this answers is not in the document]
Comment (ghtmtt) [reply source/docs/documentation_guidelines/testing_guidelines.rst]: it is not just a venv setup, you have to to install docker. I really would skip windows part here (or a Windows user can add it in a second moment, but I'm afraid things are not as smooth as in linux)

**Comment by ghtmtt** [comment]
@SrNetoChan something weird appeared in the PR you made on my repo (gh says I don't have the correct permission.....). Anyway I made a combo between your PR in my repo and all the reviews in this conversation.

File: source/docs/documentation_guidelines/testing_guidelines.rst [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    print("QGIS CRS ID:", crs.srsid())
+    print("PostGIS SRID:", crs.srid())
+
+and then adding the expected results within the ``.. testoutput:: mysnippet``::
+
+    QGIS CRS ID: 3452
+    PostGIS SRID: 4326
+
+
+How to test snippets on your local machine
```
[the remark this answers is not in the document]
Comment (ghtmtt) [reply source/docs/documentation_guidelines/testing_guidelines.rst]: why not? Skilled advanced users maybe would like to know how to test the code locally. I'm in favor of let this section, I think is more visible than in the readme file

**Comment by SrNetoChan** [comment]
@ghtmtt but you haven't pushed it yet, right? @DelazJ made some suggestions there too.

**Comment by ghtmtt** [comment]
> @ghtmtt but you haven't pushed it yet, right? @DelazJ made some suggestions there too.

nope not pushed, just took you changes and the comments of both PR and made a combo

**Comment by SrNetoChan** [comment]
Sorry for the mess, But I was getting really confused trying to doing the changes I though would be needed as suggestions.

File: source/docs/documentation_guidelines/index.rst [hunk 10]
```diff
@@ -11,3 +11,4 @@ Documentation Guidelines
    do_translations
    substitutions
    first_contribution
+   cookbook_guidelines
```
Comment (DelazJ) [inline source/docs/documentation_guidelines/index.rst]: I still think it should be moved up. It shows how to write and should be grouped with how to write instructions (more discoverable), ie
`writing rules (text or code) ---> translation rule --> substitution (a help to write) --> how to play with git/github`
```suggestion
```

File: source/docs/documentation_guidelines/index.rst [hunk 11]
```diff
@@ -11,3 +11,4 @@ Documentation Guidelines
    do_translations
```
Comment (DelazJ) [inline source/docs/documentation_guidelines/index.rst]:
```suggestion
   cookbook_guidelines
   do_translations
```

**Comment by ghtmtt** [comment]
@DelazJ ok it seems we have a small conflict on the chapter order :) let's be democratic and hear for at least another opinion. @SrNetoChan? @havatv?

**Comment by SrNetoChan** [comment]
I am ok with the chapter order @DelazJ suggested

File: source/docs/documentation_guidelines/testing_guidelines.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (SrNetoChan) [reply source/docs/documentation_guidelines/testing_guidelines.rst]: I am +0, having it here helps people understand why they need to use these test blocks. Besides, people editing the code snippets will need to try on their own machines if it's working or not.

File: source/docs/documentation_guidelines/testing_guidelines.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (DelazJ) [reply source/docs/documentation_guidelines/testing_guidelines.rst]: Never mind I misread the Sphinx docs. Sorry

File: source/docs/documentation_guidelines/testing_guidelines.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (DelazJ) [reply source/docs/documentation_guidelines/testing_guidelines.rst]: So be it.

**Comment by ghtmtt** [comment]
@SrNetoChan OK then I cannot follow anymore :) Could you please point me to the not included changes?

**Comment by SrNetoChan** [comment]
@ghtmtt sorry, I know this morning was quite intense :-) 

In my proposal I did several changes in the order of the text (growing from simple to more complex topics), I have section about test groups, and have run grammarly to check grammar issues. But please merge this, I will rebase my changes on the latest commit you have and try to create a new PR on master that can be reviewed by everyone.

**Comment by SrNetoChan** [comment]
Is this ok @DelazJ ?

**Comment by DelazJ** [comment]
Yes. Let's move that way

**Comment by ghtmtt** [comment]
done!

**Pull request merged**